Fix redundant drawing connection

ericpre opened the pull request:
### Progress of the PR
- [x] Remove redundant drawing connection to avoid unnecessary re-draw and increase plotting speed,
- [x] adjust accordingly the `spikes_removal_tool`,
- [x] fix an issue with the pointer which was not updating when removing spikes,
- [x] ready for review.

ericpre committed:
Add `_init_draw` method to `BlittedFigure`.

ericpre committed:
Fix event connection with spikes_removal_tool.
Fix bug with pointer not being updated.

ericpre committed:
Find spikes at initialisation and when changing threshold.

ericpre committed:
Remove redundant drawing connection.

francisco-dlp commented on hyperspy/drawing/image.py:
```diff
@@ -292,19 +292,7 @@ def plot(self, **kwargs):
-            self.figure.canvas.flush_events()
-        except NotImplementedError:
-            pass
+        self._init_draw()
```
Wouldn't it be better to simply call ``self.figure.canvas.draw()`` instead?

francisco-dlp commented on hyperspy/drawing/signal1d.py:
```diff
@@ -125,6 +125,7 @@ def plot(self):
                 # tight_layout is a bit brittle, we do this just in case it
                 # complains
                 pass
+        self._init_draw()
```
Wouldn't it be better to simply call ``self.figure.canvas.draw()`` instead?

ericpre replied:
Yes, indeed.

francisco-dlp commented on hyperspy/drawing/figure.py:
```diff
@@ -58,6 +58,22 @@ def create_figure(self, **kwargs):
             self._draw_event_cid = self.figure.canvas.mpl_connect(
                 'draw_event', self._on_blit_draw)
 
+    def _init_draw(self):
```
This may be unnecessary, see below.

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -873,10 +873,12 @@ def __init__(self, signal, navigation_mask=None, signal_mask=None):
                 self.noise_type = "shot noise"
         else:
             self.noise_type = "shot noise"
+        self.find()
```
Why calling ``find`` here?

ericpre replied:
I am to going it to revert it because it may not be desirable if the threshold value is too high in case of large dataset since it can take too long to perform... The idea is to save the user a click on the "find" button at the start of the spikes_removal_tool, but it may be a wrong good idea! Is it what you are concerned about?

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -873,10 +873,12 @@ def __init__(self, signal, navigation_mask=None, signal_mask=None):
     def _threshold_changed(self, old, new):
         self.index = 0
         self.update_plot()
+        self.find()
```
and here?

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -950,6 +952,8 @@ def find(self, back=False):
                     minimum),
                 self.signal.axes_manager.signal_axes[0].index2value(
                     maximum))
+            self.signal.axes_manager.indices = self.coordinates[self.index]
```
Why are these two new lines necessay?

ericpre replied:
Yes one of the two is not necessary, some left over... I moved the second line at a more appropriate spot in the `update_plot()` method and this one is indeed necessary, because the pointer was not updating! Possibly since the split of the UI.

francisco-dlp commented on hyperspy/tests/signal/test_tools.py:
```diff
@@ -869,11 +869,8 @@ def test_spikes_removal_tool():
 
     sr = SpikesRemoval(s)
     sr.threshold = 1.5
-    sr.find()
```
It seems that the spikes removal tool is now behaving differently, when it shouldn't.

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -1005,9 +1009,11 @@ def create_interpolation_line(self):
             color='blue',
             type='line')
-        self.signal._plot.signal_plot.add_line(self.interpolated_line)
+        self.signal._plot.signal_plot.add_line(self.interpolated_line,
```
why shouldn't this line be updated automatically?

francisco-dlp commented on hyperspy/drawing/signal1d.py:
```diff
@@ -87,6 +87,11 @@ def add_line(self, line, ax='left'):
             line.sf_lines = self.right_ax_lines
             if line.axes_manager is None:
                 line.axes_manager = self.right_axes_manager
+        if connect_event:
```
If this is finally not necessary for this PR (to be discussed) I would suggest to remove it.

ericpre replied:
There are situations, where you don't want to connect the second line (spikes_removal_tool, model line , etc, need to check) to avoid redrawing a second time, the same thing.

ericpre replied:
Do you think it is worth to look for places, when this may be the case and fixes it accordingly?

francisco-dlp replied:
I think that it's better to stick to the topic of this particular PR. I think that this is a good idea in general as it decreases the time needed to review individual PRs and it makes it easier to find a short title that describes the changes.

francisco-dlp replied:
If this feature is gonna stay I would suggest to rename the keyword to something more explicit such as ``update_on_navigation`` or something of this sort.

ericpre committed:
Address @francisco-dlp's review comments:
- remove unnecessary line (setting axes_manager indices),
- simplify drawing at the end of the plot method of `ImagePlot` and `Signal1DFigure`,
- revert back the wrong good idea of finding automatically spikes when updating threshold value and at initialisation of the `spikes_removal_tool`: on some dataset, it may took long and hold the UI.

francisco-dlp commented on hyperspy/tests/signal/test_tools.py:
```diff
@@ -872,7 +872,6 @@ def test_spikes_removal_tool():
     sr.find()
     assert s.axes_manager.indices == (0, 1)
     sr.threshold = 0.5
-    assert s.axes_manager.indices == (0, 0)
```
This was testing that changing the threshold sets the indices back to zero, why removing it?

francisco-dlp commented on hyperspy/drawing/signal1d.py:
```diff
@@ -310,7 +307,6 @@ def plot(self, data=1):
                                      fontsize=12,
                                      color=self.line.get_color(),
                                      animated=self.ax.figure.canvas.supports_blit)
-        self.ax.figure.canvas.draw_idle()
```
Doesn't removing this imply that the line is not drawn when calling the ``plot`` method? If yes, why would that be a good idea?

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -1005,9 +1005,11 @@ def create_interpolation_line(self):
+                                               connect_event=False)
         self.interpolated_line.autoscale = False
         self.interpolated_line.plot()
+        self.interpolated_line.update()
```
Why do we need to call ``update`` here if the line data haven't changed?

ericpre committed:
Address @francisco-dlp's review comments take2: revert back a few things which were not really sensible.

ericpre commented:
Ready for review again: indeed, there were a few things not very sensible!

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -960,6 +959,7 @@ def update_plot(self):
         self.reset_span_selector()
         self.update_spectrum_line()
         if len(self.coordinates) > 1:
+            self.signal._plot.pointer.indices = self.coordinates[self.index]
```
This is actually a workaround for what may be a minor bug: when ``_auto_update_plot`` is switched from ``False`` to ``True`` the pointer's indices should be set  to the current ones and the patch redrawn, probably around https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/drawing/mpl_hse.py#L55

ericpre replied:
Yes, `auto_update_plot=True` will update the pointer position and redraw the pointer nicely but it will do it for each index and slows down significantly finding spikes, since a indices change of the axes manager will redraw the patch position! So it seems that it is working as expected but not what we want.
`_update_patch_position()` is also working fine but since the actual pointer indices haven't changed, the same thing is redrawn. The only way that I see here is setting the pointer indices manually and changing `_update_patch_position()` to `_on_navigate(self.signal.axes_manager)` does the job here. Actually, just using the pointer position or indices properties would also do, since it trigger `_pos_changed` event and call `_update_patch_position`.

francisco-dlp commented on hyperspy/signal_tools.py:
```diff
@@ -1005,7 +1005,8 @@ def create_interpolation_line(self):
             type='line')
-        self.signal._plot.signal_plot.add_line(self.interpolated_line)
+        self.signal._plot.signal_plot.add_line(self.interpolated_line,
+                                               update_on_navigation=False)
```
I think that ``update_on_navigation`` may not be needed as setting its ``auto_update`` attribute to ``False`` achieves the same effect and, ideally, there should only be one way of achieving the effect.

ericpre replied:
Yes, this is better, with a caveat that we need to ignore the `auto_update=False` on the `span_selector_changed` callback.

francisco-dlp commented on hyperspy/drawing/signal1d.py:
```diff
@@ -87,6 +87,11 @@ def add_line(self, line, ax='left'):
             line.sf_lines = self.right_ax_lines
             if line.axes_manager is None:
                 line.axes_manager = self.right_axes_manager
+        if update_on_navigation:
```
I think that this may not be needed, see below.

ericpre replied:
Done.

ericpre committed:
Address @francisco-dlp's review comments take3!
- Use `auto_update` setting of `Signal1DLine` in favour of introducing a new argument to the `add_line` method. Create a `ignore_auto_update` argument for the `update` method of `Signal1DLine`.
- Use `_on_navigate()` in favour of `_update_patch_position()` to update the pointer position.

francisco-dlp committed:
Fix line.update was not updating auto_update=False

Commit:
Automatic style corrections courtesy of autopep8

ericpre committed:
Merge pull request #6 from francisco-dlp/FIX_redundant_drawing_connection

Fix redundant drawing connection
